src/array/cpu/spmat_op_impl.cc

```diff
@@ -0,0 +1,635 @@
+    std::copy(data + indptr_data[rid], data + indptr_data[rid + 1],
+              ret_data + ret_indptr_data[i]);
+  }
+  return ret;
+}
+
+template CSRMatrix CSRSliceRows<kDLCPU, int32_t, int32_t>(CSRMatrix , NDArray);
+template CSRMatrix CSRSliceRows<kDLCPU, int64_t, int64_t>(CSRMatrix , NDArray);
+
+///////////////////////////// CSRSliceMatrix /////////////////////////////
+
+template <DLDeviceType XPU, typename IdType, typename DType>
+CSRMatrix CSRSliceMatrix(CSRMatrix csr, runtime::NDArray rows, runtime::NDArray cols) {
+  CHECK(CSRHasData(csr)) << "missing data array.";
+  IdHashMap<IdType> hashmap(cols);
+  const int64_t new_nrows = rows->shape[0];
+  const int64_t new_ncols = cols->shape[0];
+  const IdType* rows_data = static_cast<IdType*>(rows->data);
+
+  const IdType* indptr_data = static_cast<IdType*>(csr.indptr->data);
+  const IdType* indices_data = static_cast<IdType*>(csr.indices->data);
+  const DType* data = static_cast<DType*>(csr.data->data);
+
+  std::vector<IdType> sub_indptr, sub_indices;
+  std::vector<DType> sub_data;
+  sub_indptr.resize(new_nrows + 1, 0);
+  const IdType kInvalidId = new_ncols + 1;
+  for (int64_t i = 0; i < new_nrows; ++i) {
+    // NOTE: newi == i
+    const IdType oldi = rows_data[i];
+    CHECK(oldi >= 0 && oldi < csr.num_rows) << "Invalid row index: " << oldi;
+    for (IdType p = indptr_data[oldi]; p < indptr_data[oldi + 1]; ++p) {
+      const IdType oldj = indices_data[p];
+      const IdType newj = hashmap.Map(oldj, kInvalidId);
+      if (newj != kInvalidId) {
+        ++sub_indptr[i];
+        sub_indices.push_back(newj);
+        sub_data.push_back(data[p]);
+      }
+    }
+  }
+
+  // cumsum sub_indptr
+  for (int64_t i = 0, cumsum = 0; i < new_nrows; ++i) {
+    const IdType temp = sub_indptr[i];
+    sub_indptr[i] = cumsum;
+    cumsum += temp;
+  }
+  sub_indptr[new_nrows] = sub_indices.size();
+
+  const int64_t nnz = sub_data.size();
+  NDArray sub_data_arr = NDArray::Empty({nnz}, csr.data->dtype, csr.data->ctx);
+  DType* ptr = static_cast<DType*>(sub_data_arr->data);
+  std::copy(sub_data.begin(), sub_data.end(), ptr);
+  return CSRMatrix{new_nrows, new_ncols,
+    VecToIdArray(sub_indptr), VecToIdArray(sub_indices), sub_data_arr};
+}
+
+template CSRMatrix CSRSliceMatrix<kDLCPU, int32_t, int32_t>(
+    CSRMatrix csr, runtime::NDArray rows, runtime::NDArray cols);
+template CSRMatrix CSRSliceMatrix<kDLCPU, int64_t, int64_t>(
+    CSRMatrix csr, runtime::NDArray rows, runtime::NDArray cols);
+
+///////////////////////////// COOHasDuplicate /////////////////////////////
+
+template <DLDeviceType XPU, typename IdType>
+bool COOHasDuplicate(COOMatrix coo) {
+  std::unordered_set<std::pair<IdType, IdType>, PairHash> hashmap;
+  const IdType* src_data = static_cast<IdType*>(coo.row->data);
+  const IdType* dst_data = static_cast<IdType*>(coo.col->data);
+  const auto nnz = coo.row->shape[0];
+  for (IdType eid = 0; eid < nnz; ++eid) {
```

Comment:
this is definitely not right. 2B x 2B matrix can have more than 2B edges. we should use int64 here. a general suggest: try using int64 as much as possible for index in for loops.